core/src/main/java/org/jdbi/v3/core/SqlArrayArgumentFactory.java

```diff
@@ -23,19 +23,31 @@
 
 class SqlArrayArgumentFactory implements ArgumentFactory {
     @Override
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({ "unchecked", "rawtypes" })
     public Optional<Argument> build(Type type, Object value, StatementContext ctx) {
         Class<?> erasedType = GenericTypes.getErasedType(type);
         if (erasedType.isArray()) {
             Class<?> elementType = erasedType.getComponentType();
             return ctx.findArrayTypeFor(elementType)
-                    .map(arrayType -> new SqlArrayArgument(arrayType, value));
+                    .map(arrayType -> new SqlArrayArgument<>(arrayType, value));
         }
-        if (List.class.isAssignableFrom(erasedType)) {
-            return GenericTypes.findGenericParameter(type, List.class)
+        if (Collection.class.isAssignableFrom(erasedType)) {
+            return guessElementType(type, (Collection<?>) value)
                     .flatMap(ctx::findArrayTypeFor)
-                    .map(arrayType -> new SqlArrayArgument(arrayType, (List) value));
+                    .map(arrayType -> new SqlArrayArgument<>(arrayType, (Collection) value));
         }
         return Optional.empty();
     }
+
+    private Optional<? extends Type> guessElementType(Type type, Collection<?> value) {
```

Comment:
Right now if you `bind("ids", new ArrayList<>(1, 2, 3))`, `findGenericParameter` returns `empty` because the runtime type is erased.  Guessing based on the elements is kinda shitty, and we fall back to `Object`, but I'm not sure what else to do.  (You can be explicit with `bindByType` but asking our users to do that for every array sucks.)